File: src/main/java/uk/gov/justice/digital/hmpps/riskprofiler/dao/DataRepositoryFactory.java
package uk.gov.justice.digital.hmpps.riskprofiler.dao;

import org.springframework.stereotype.Component;
import uk.gov.justice.digital.hmpps.riskprofiler.datasourcemodel.*;

import java.util.List;

@Component
public class DataRepositoryFactory {

    private final DataRepository<Ocgm> ocgmRepository;
    private final DataRepository<Ocg> ocgRepository;
    private final DataRepository<PathFinder> pathfinderRepository;
    private final DataRepository<Pras> prasRepository;
    private final DataRepository<Viper> viperRepository;

    public DataRepositoryFactory(OcgmRepository ocgmRepository, OcgRepository ocgRepository, PathfinderRepository pathfinderRepository, PrasRepository prasRepository, ViperRepository viperRepository) {
        this.ocgmRepository = ocgmRepository;
        this.ocgRepository = ocgRepository;
        this.pathfinderRepository = pathfinderRepository;
        this.prasRepository = prasRepository;
        this.viperRepository = viperRepository;
    }

    @SuppressWarnings("unchecked")
    public <T extends RiskDataSet> DataRepository<T> getRepository(Class<T> type) {

        DataRepository<T> repository = null;

        switch (FileType.byDataSet(type)) {

            case PRAS:
                return (DataRepository<T>)prasRepository;

            case OCGM:
                return (DataRepository<T>)ocgmRepository;

            case OCG:
                return (DataRepository<T>)ocgRepository;

            case PATHFINDER:
                return (DataRepository<T>)pathfinderRepository;

            case VIPER:
                return (DataRepository<T>)viperRepository;
        }

        return repository;
    }

    public List<DataRepository<? extends RiskDataSet>> getRepositories() {
        return List.of(
                getRepository(Pras.class),
                getRepository(Viper.class),
                getRepository(Ocg.class),
                getRepository(Ocgm.class),
                getRepository(PathFinder.class));
    }
}

File: src/main/kotlin/uk/gov/justice/digital/hmpps/riskprofiler/dao/DataRepositoryFactory.kt
package uk.gov.justice.digital.hmpps.riskprofiler.dao

import org.springframework.stereotype.Component
import uk.gov.justice.digital.hmpps.riskprofiler.datasourcemodel.FileType
import uk.gov.justice.digital.hmpps.riskprofiler.datasourcemodel.FileType.Companion.byDataSet
import uk.gov.justice.digital.hmpps.riskprofiler.datasourcemodel.Ocg
import uk.gov.justice.digital.hmpps.riskprofiler.datasourcemodel.OcgmList
import uk.gov.justice.digital.hmpps.riskprofiler.datasourcemodel.Pras
import uk.gov.justice.digital.hmpps.riskprofiler.datasourcemodel.RiskDataSet
import uk.gov.justice.digital.hmpps.riskprofiler.datasourcemodel.Viper

@Component("dataRepositoryFactory")
class DataRepositoryFactory(
  ocgmRepository: OcgmRepository,
  ocgRepository: OcgRepository,
  prasRepository: PrasRepository,
  viperRepository: ViperRepository,
) {
  private val ocgmRepository: DataRepository<OcgmList>
  private val ocgRepository: DataRepository<Ocg>
  private val prasRepository: DataRepository<Pras>
  private val viperRepository: DataRepository<Viper>
  fun <T : RiskDataSet> getRepository(type: Class<T>): DataRepository<T> {
    // val repository: DataRepository<T> = null
    return when (byDataSet(type)) {
      FileType.PRAS -> prasRepository as DataRepository<T>
      FileType.OCGM -> ocgmRepository as DataRepository<T>
      FileType.OCG -> ocgRepository as DataRepository<T>
      FileType.VIPER -> viperRepository as DataRepository<T>
    }
    // return repository
  }

  fun getRepositories(): List<DataRepository<out RiskDataSet>> {
    return listOf(
      getRepository(Pras::class.java),
      getRepository(Viper::class.java),
      getRepository(Ocg::class.java),
      getRepository(OcgmList::class.java),
    )
  }

  init {
    this.ocgmRepository = ocgmRepository
    this.ocgRepository = ocgRepository
    this.prasRepository = prasRepository
    this.viperRepository = viperRepository
  }
}
